File: src/types/BlockTag.js
/**
 * @fileoverview Block tag enum for referencing specific blocks
 *
 * Uses @tevm/primitives for type compatibility across the ecosystem.
 */

/**
 * Block tags for referencing specific blocks.
 *
 * - `earliest`: The lowest numbered block the client has available
 * - `finalized`: The most recent crypto-economically secure block, cannot be re-orged outside of manual intervention driven by community coordination
 * - `safe`: The most recent block that is safe from re-orgs under honest majority and certain synchronicity assumptions
 * - `latest`: The most recent block in the canonical chain observed by the client, this block may be re-orged out of the canonical chain even under healthy/normal conditions
 * - `pending`: A sample next block built by the client on top of `latest` and containing the set of transactions usually taken from local mempool
 *
 * **Note**: Compatible with `@tevm/primitives/ethereum-types` BlockTag type.
 *
 * @see https://github.com/ethereum/execution-apis
 */
export const BlockTag = {
  earliest: 'earliest',
  finalized: 'finalized',
  safe: 'safe',
  latest: 'latest',
  pending: 'pending',
}

/**
 * Type representing a block tag value.
 *
 * @typedef {(typeof BlockTag)[keyof typeof BlockTag]} BlockTag
 */

File: src/types/BlockTag.zig
const std = @import("std");

/// Block tag for referencing specific blocks
///
/// Follows @tevm/primitives conventions for BlockTag type.
///
/// - `earliest`: The lowest numbered block the client has available
/// - `finalized`: The most recent crypto-economically secure block
/// - `safe`: The most recent block that is safe from re-orgs
/// - `latest`: The most recent block in the canonical chain
/// - `pending`: A sample next block built on top of `latest`
pub const BlockTag = enum {
    earliest,
    finalized,
    safe,
    latest,
    pending,

    pub fn toString(self: BlockTag) []const u8 {
        return switch (self) {
            .earliest => "earliest",
            .finalized => "finalized",
            .safe => "safe",
            .latest => "latest",
            .pending => "pending",
        };
    }

    pub fn fromString(str: []const u8) !BlockTag {
        if (std.mem.eql(u8, str, "earliest")) return .earliest;
        if (std.mem.eql(u8, str, "finalized")) return .finalized;
        if (std.mem.eql(u8, str, "safe")) return .safe;
        if (std.mem.eql(u8, str, "latest")) return .latest;
        if (std.mem.eql(u8, str, "pending")) return .pending;
        return error.InvalidBlockTag;
    }

    /// Serialize to JSON using Zig 0.15 API
    pub fn jsonStringify(self: BlockTag, jws: *std.json.Stringify) !void {
        try jws.write(self.toString());
    }

    /// Deserialize from JSON using Zig 0.15 API
    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !BlockTag {
        _ = allocator;
        _ = options;
        if (source != .string) return error.UnexpectedToken;
        return try fromString(source.string);
    }
};
